### PR TITLE
vfc_unmap_map test script

### PR DESCRIPTION
**Pull request by Naresh-ibm**
Adding test script for vfc_unmap_map test. skeleton is all done.
ignore cosmetic changes. review the logic and flow of the script run.
And also point out if any major changes is required for the script to get merge.

Signed-off-by: Naresh Bannoth <nbannoth@in.ibm.com>

### CODE REVIEW

**Comment by narasimhan-v** [comment]
@Naresh-ibm Why can't this be part of https://github.com/avocado-framework-tests/avocado-misc-tests/blob/master/io/disk/vfc-tests/vfc_dlpar.py ?

File: io/disk/vfc-tests/vfc_unmap_map.py [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+                self.log.info("complete list at vios : %s" % self.vfc_dic)
+                self.vfc_dic["vfchost"] = vfchost
+                self.log.info("complete list vfchost : %s" % self.vfc_dic)
+                self.vfc_dic["fcs"] = self.get_fcs_name(vfchost, vios)
+                self.log.info("complete list at fcs : %s" % self.vfc_dic)
+                self.vfc_dic["vfc_client"] = self.get_vfc_client(vfchost, vios)
+                self.log.info("complete list at vfc_client: %s" % self.vfc_dic)
+                self.vfc_dic["paths"] = self.get_paths(self.vfc_dic["vfc_client"])
+                self.log.info("complete list at paths: %s" % self.vfc_dic)
+                self.dic_list.append(self.vfc_dic)
```
Comment (narasimhan-v) [inline io/disk/vfc-tests/vfc_unmap_map.py]: local scope for `vfc_dic` is fine.

File: io/disk/vfc-tests/vfc_unmap_map.py [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+        returns the linux_name of corresponding drc_name
+        '''
+        cmd = 'viosvrcmd -m %s -p %s -c "lsmap -all -npiv -field Name \
+               fc -fmt :"' % (self.server, vios_name)
+        output = self.run_command(cmd)
+        self.log.info("output value : %s" % output)
+        for line in output:
+            if vfchost in line:
+                return line.split(":")[-1]
+        return None
```
Comment (narasimhan-v) [inline io/disk/vfc-tests/vfc_unmap_map.py]: `return ''`

File: io/disk/vfc-tests/vfc_unmap_map.py [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+        returns the linux_name of corresponding drc_name
+        '''
+        cmd = 'viosvrcmd -m %s -p %s -c "lsmap -all -npiv -field  name\
+               vfcclient -fmt :"' % (self.server, vios_name)
+        output = self.run_command(cmd)
+        self.log.info("output value : %s" % output)
+        for line in output:
+            if vfchost in line:
+                return line.split(":")[-1]
+        return None
```
Comment (narasimhan-v) [inline io/disk/vfc-tests/vfc_unmap_map.py]: `return ''`

File: io/disk/vfc-tests/vfc_unmap_map.py [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+        returns the linux_name of corresponding drc_name
+        '''
+        cmd = 'viosvrcmd -m %s -p %s -c "lsmap -all -npiv -field  name\
+               Status -fmt :"' % (self.server, vios_name)
+        output = self.run_command(cmd)
+        self.log.info("output value : %s" % output)
+        for line in output:
+            if vfchost in line:
+                return line.split(":")[-1]
+        return None
```
Comment (narasimhan-v) [inline io/disk/vfc-tests/vfc_unmap_map.py]: `return ''`

File: io/disk/vfc-tests/vfc_unmap_map.py [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+                paths.append(line.split("/")[-1])
+        return paths
+
+    def vfc_client_status_verify(self, operation, vfc_client):
+        '''
+        Returns the WWPNs of give client slot number
+        '''
+        err_host = []
+        cmd = 'cat /sys/class/fc_host/host2/port_state'
+        status = process.system_output(cmd).decode('utf-8')
```
Comment (narasimhan-v) [inline io/disk/vfc-tests/vfc_unmap_map.py]: please use `genio.read_file()`

File: io/disk/vfc-tests/vfc_unmap_map.py [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+            cmd = 'vfcmap -vadapter %s -fcp %s' \
+                   % (vfc_dic["vfchost"], vfc_dic["fcs_name"])
+        else:
+            cmd = 'vfcmap -vadapter %s -fcp %s' % (vfc_dic["vfchost"])
+        try:
+            self.run_command(cmd)
+        except CommandFailed as cmdfail:
+            self.log.debug(str(cmdfail))
+            self.fail("vfchost %s operation failed" % operation)
+        time.sleep(self.opp_sleep_time)
```
Comment (narasimhan-v) [inline io/disk/vfc-tests/vfc_unmap_map.py]: why do we need sleep here ?

File: io/disk/vfc-tests/vfc_unmap_map.py [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+        return ''
+
+    def test(self):
+        '''
+        Remove and add vfc interfaces Dynamically from HMC
+        '''
+        for _ in range(self.count):
+            for vfc_dic in self.dic_list:
+                self.vfchost_map_unmap("unmap", vfc_dic)
+                time.sleep(self.opp_sleep_time)
```
Comment (narasimhan-v) [inline io/disk/vfc-tests/vfc_unmap_map.py]: why do we need sleep here ?
Also, why sleep only after unmap and not after map ?

Comment (Naresh-ibm) [reply io/disk/vfc-tests/vfc_unmap_map.py]: need to add the sleep even after map too. we need the sleep for sure but value can be reduced. will do

File: io/disk/vfc-tests/vfc_unmap_map.py (hunk 6 above)
Comment (Naresh-ibm) [reply io/disk/vfc-tests/vfc_unmap_map.py]: before verification need some sleep time for changes to  reflections like status of different elements. so this is required.

File: io/disk/vfc-tests/vfc_unmap_map.py (hunk 5 above)
Comment (Naresh-ibm) [reply io/disk/vfc-tests/vfc_unmap_map.py]: ok

File: io/disk/vfc-tests/vfc_unmap_map.py (hunk 4 above)
Comment (Naresh-ibm) [reply io/disk/vfc-tests/vfc_unmap_map.py]: why not None?

File: io/disk/vfc-tests/vfc_unmap_map.py (hunk 1 above)
Comment (Naresh-ibm) [reply io/disk/vfc-tests/vfc_unmap_map.py]: ok, will change

File: io/disk/vfc-tests/vfc_unmap_map.py (hunk 4 above)
Comment (narasimhan-v) [reply io/disk/vfc-tests/vfc_unmap_map.py]: all returns in a function should return same type

File: io/disk/vfc-tests/vfc_unmap_map.py (hunk 6 above)
Comment (narasimhan-v) [reply io/disk/vfc-tests/vfc_unmap_map.py]: can't we have some check in sysfs instead of sleep ?

File: io/disk/vfc-tests/vfc_unmap_map.py (hunk 7 above)
Comment (narasimhan-v) [reply io/disk/vfc-tests/vfc_unmap_map.py]: I would say, `map()` should make the system stable so that you don't need sleep here.
And in `map()`, you should find a way to make it stable without sleep. sleep is not a good way to handle things, it is only a workaround.

Comment (Naresh-ibm) [reply io/disk/vfc-tests/vfc_unmap_map.py]: the status change will different amount of time in different distros
I had a thought of using wait functions but the status may change or may not chage due to some issue. hence again we have to provide the wait time there too, which as good as sleep time only. and I have one more functionality check here (fast_io_fail_tmo anddev_loss_tmo attribute check) for which I have to give sleep time for sure. so having sleep time in common will make the code simple. hence used.

So as per my knowledge, we have two way
1. use wait.wait_for(mpath_svc.status, timeout=10) and use sleep time for above attribute functionality check
2. use the sleep_time for both

let me know which you feel is better

for option 1 I have to change the verification functions to TRUE or FALSE to and need changes at multiple places. and if needed , have to write a separate function to handle it.

File: io/disk/vfc-tests/vfc_unmap_map.py (hunk 4 above)
Comment (Naresh-ibm) [reply io/disk/vfc-tests/vfc_unmap_map.py]: I have used None for all returns right?

File: io/disk/vfc-tests/vfc_unmap_map.py (hunk 7 above)
Comment (narasimhan-v) [reply io/disk/vfc-tests/vfc_unmap_map.py]: @Naresh-ibm option sounds better

File: io/disk/vfc-tests/vfc_unmap_map.py [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+            file_name = '/sys/class/fc_host/%s/port_state' % vfc_client
+            status = genio.read_file(file_name).strip("\n")
+            if operation == "map":
+                if status == 'Online':
+                    return True
+                return False
+            elif operation == 'unmap':
+                if status == 'Linkdown':
+                    return True
+                return False
```
Comment (narasimhan-v) [inline io/disk/vfc-tests/vfc_unmap_map.py]: Why is this needed inside the function vfc_client_status_verify() ?

File: io/disk/vfc-tests/vfc_unmap_map.py [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+        def status_check():
+            vfchost_status = self.get_vfchost_status(vfchost, vios_name)
+            if operation == "map":
+                if vfchost_status == 'LOGGED_IN':
+                    return True
+                return False
+            elif operation == "unmap":
+                if vfchost_status == 'NOT_LOGGED_IN':
+                    return True
+                return False
```
Comment (narasimhan-v) [inline io/disk/vfc-tests/vfc_unmap_map.py]: Suggestion:
```
            if operation == "map":
                if vfchost_status == 'LOGGED_IN':
                    return True
            elif operation == "unmap":
                if vfchost_status == 'NOT_LOGGED_IN':
                    return True
            return False
```

Comment (Naresh-ibm) [reply io/disk/vfc-tests/vfc_unmap_map.py]: ok

File: io/disk/vfc-tests/vfc_unmap_map.py (hunk 8 above)
Comment (Naresh-ibm) [reply io/disk/vfc-tests/vfc_unmap_map.py]: this is to avoid the waiting time that we use to have with time.sleep

Comment (narasimhan-v) [reply io/disk/vfc-tests/vfc_unmap_map.py]: @Naresh-ibm My question was: "Why is this needed inside the function vfc_client_status_verify() ?"

**Comment by narasimhan-v** [comment]
@Naresh-ibm Can you please check if this is the latest version of the code ?

**Comment by Naresh-ibm** [comment]
> @Naresh-ibm Can you please check if this is the latest version of the code ?

@narasimhan-v  yes this is the latest code

**Comment by narasimhan-v** [comment]
@Naresh-ibm I see the copyright is back to 2019, and the sleep time still at 120s. Please correct those.

File: io/disk/vfc-tests/vfc_unmap_map.py [hunk 10]
```diff
@@ -0,0 +1,415 @@
...
+
+    def test_undefine_define(self):
+        '''
+        Undefin and define the vfchost from vios
+        '''
+        self.err_mesg = []
+        for _ in range(self.count):
+            for vfc_dic in self.dic_list:
+                self.vfchost_define_undefine("undefine", vfc_dic)
+                time.sleep(self.opp_sleep_time)
```
Comment (narasimhan-v) [inline io/disk/vfc-tests/vfc_unmap_map.py]: # TODO: <reason>

Comment (narasimhan-v) [reply io/disk/vfc-tests/vfc_unmap_map.py]: @Naresh-ibm can this be taken care of ?

**Comment by Naresh-ibm** [comment]
[root@ltcfleet2-lp10 ~]# avocado run unmap_map.py -m unmap_map.yaml 
Fetching assets from unmap_map.py:VirtualFC.test_unmap_map
Error running method "pre_tests" of plugin "fetchasset": 'Attribute' object has no attribute 'id'
JOB ID     : 515a439f34e67b459d4db6db127e254b2ed8b19f
JOB LOG    : /root/avocado-fvt-wrapper/results/job-2020-02-24T15.59-515a439/job.log
 (1/3) unmap_map.py:VirtualFC.test_unmap_map;run-6107: FAIL: test failed due to folowing reasons:['after unmap host2 staus change                                   failed', 'after unmap host3 staus change                                   failed', 'after unmap host4 staus change                                   fa... (1078.10 s)
 (2/3) unmap_map.py:VirtualFC.test_undefine_define;run-6107: FAIL: test failed due to folowing reasons:['after undefine host2 staus change                                   failed', 'after undefine path sdb status did not                                       changed', 'after undefine path sdc status did not             ... (1921.64 s)
 (3/3) unmap_map.py:VirtualFC.test_create_delete_vfchost;run-6107: PASS (298.74 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 2 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 3298.89 s
JOB HTML   : /root/avocado-fvt-wrapper/results/job-2020-02-24T15.59-515a439/results.html
[root@ltcfleet2-lp10 ~]#

**Comment by Naresh-ibm** [comment]
test unamp_map and define_undefine are expected failures. there is a real issue and it has been reported already.
so ignore the failure in those cases.

**Comment by Naresh-ibm** [comment]
not able to upload the tar file. placed it in the box folder at 
https://ibm.ent.box.com/folder/104632712129.

let me know if you have any problem in accessing the log file

**Comment by narasimhan-v** [comment]
[virtual_fc_job.log.zip](https://github.com/avocado-framework-tests/avocado-misc-tests/files/4249975/virtual_fc_job.log.zip)

File: io/disk/vfc-tests/vfc_unmap_map.py (hunk 10 above)
Comment (narasimhan-v) [reply io/disk/vfc-tests/vfc_unmap_map.py]: @Naresh-ibm Can this be taken care of ?

File: io/disk/vfc-tests/virtual_fc.py [hunk 11]
```diff
@@ -0,0 +1,531 @@
...
+    def test_unmap_map(self):
+        '''
+        Remove and add vfc interfaces Dynamically from HMC
+        '''
+        for _ in range(self.count):
+            for vfc_dic in self.dic_list:
+                self.vfchost_map_unmap("unmap", vfc_dic)
+                time.sleep(self.opp_sleep_time)
+                self.vfchost_map_unmap("map", vfc_dic)
+                # sleeping between the operation. can be enhanced in future.
```
Comment (narasimhan-v) [inline io/disk/vfc-tests/virtual_fc.py]: @Naresh-ibm We need the reason for sleep, and why it is the only possible solution now.

File: io/disk/vfc-tests/virtual_fc.py [hunk 12]
```diff
@@ -0,0 +1,531 @@
...
+        return ''
+
+    def test_unmap_map(self):
+        '''
+        Remove and add vfc interfaces Dynamically from HMC
+        '''
+        for _ in range(self.count):
+            for vfc_dic in self.dic_list:
+                self.vfchost_map_unmap("unmap", vfc_dic)
+                time.sleep(self.opp_sleep_time)
```
Comment (narasimhan-v) [inline io/disk/vfc-tests/virtual_fc.py]: here also

File: io/disk/vfc-tests/virtual_fc.py [hunk 13]
```diff
@@ -0,0 +1,531 @@
...
+
+    def test_undefine_define(self):
+        '''
+        Undefin and define the vfchost from vios
+        '''
+        self.err_mesg = []
+        for _ in range(self.count):
+            for vfc_dic in self.dic_list:
+                self.vfchost_define_undefine("undefine", vfc_dic)
+                time.sleep(self.opp_sleep_time)
```
Comment (narasimhan-v) [inline io/disk/vfc-tests/virtual_fc.py]: here as well

File: io/disk/vfc-tests/virtual_fc.py [hunk 14]
```diff
@@ -0,0 +1,531 @@
...
+                self.vfchost_define_undefine("define", vfc_dic)
+                # sleep time between operations, can be enhanced in future.
```
Comment (narasimhan-v) [inline io/disk/vfc-tests/virtual_fc.py]: here as well
